components/relevancy/src/lib.rs

```diff
@@ -0,0 +1,80 @@
+
+//! Proposed API for the relevancy component (validation phase)
+//!
+//! The goal here is to allow us to validate that we can reliably detect user interests from
+//! history data, without spending too much time building the API out.  There's some hand-waving
+//! towards how we would use this data to rank search results, but we don't need to come to a final
+//! decision on that yet.
+
+/// List of possible interests for a domain.  Each domain would have exactly one of these.  If we
+/// can't calculate an interest based on the classifier data, then we put it in `Unknown`
+#[derive(uniffi::Enum)]
+pub enum Interest {
+    Education,
+    Gaming,
+    Sports,
+    Unknown,
+}
+
+/// An interest combined with a relevancy score
+///
+/// Scores will be [0-10].  Consumers would use this score to alter the order of their suggestions.
+///
+/// For example, consumers may assign a default score of 5 to sponsored suggestions and 0 to
+/// Wikipedia suggestions.  By default, the sponsored suggestions will show up over Wikipedia
+/// suggestions, but if a Wikipedia suggestion had a interest that was highly relevant for the
+/// user, that suggestion would show up first instead.
+#[derive(uniffi::Record)]
+pub struct InterestScore {
+    interest: Interest,
+    score: u8,
+}
+
+/// Top-level class that ingests and stores relevancy data.
+///
+/// For the validation phase:
+///   - All data will be stored in memory.
+///   - Consumers will call `ingest` then `get_top_categories` and store the results in a Glean
+///     distribution (see `metrics.yaml).  Our main focus will be to monitor these metrics in order
+///     to determine if the relevancy scores would be strong enough to affect search rankings.
+///   - If possible, I'd also like to have make "about::relevancy" where users can go to see the
+///     results themselves.  This way we can pair the Glean metrics with asking people if they
+///     think the interest rankings seem correct to them.
+#[derive(uniffi::Object)]
+pub struct RelavancyStore {
+}
+
+#[uniffi::export]
+impl RelavancyStore {
+    /// Ingest relevancy data from remote settings
+    ///
+    /// This will download the classifier data from remote settings, then process the top N domains
+    /// by frecency against that data.  Each domain will be classified into 1 interest, which may be
+    /// `Interest::Unknown`.
+    ///
+    /// Based on that, we will build a interest vector.  There will be one dimension for each
+    /// interest where the value is the number of domains that were classified with that interest.
+    pub fn ingest(&self) {
+        todo!()
+    }
+
+    /// Get the highest 3 interests by score.
+    ///
+    /// To calculate the scores:
+    ///   - Take the interest vector from `ingest()`
+    ///   - Find the maximum 3 values, by raw numbers (excluding Interest::Unknown)
+    ///   - Normalize the value by dividing by the norm of the vector
+    ///      - The norm is `sqrt(sum([n*n for n in vec]))`
```

Comment:
If I understood the wikipedia article correctly, this is taking the [cosine similarity](https://en.wikipedia.org/wiki/Cosine_similarity) between the entire interest vector and a unit vector in the direction that one interest.

Reply:
Ah, cosine similarities! Though I am a little confused that we normally use cosine similarities to compare two vectors, looks like we only one vector here, correct? At the first glance, I thought we wanted to use that to calculate the relevance score for some categorized content against user's interests (so that we have two vectors).

There are a couple of ways to normalize a category hit vector into scores, such as arithmetic means or [soft-max](https://en.wikipedia.org/wiki/Softmax_function).

Reply:
This comes from an assumption that consumers would only want to specify a single interest for a suggestion, so we would have a vector where one element was `1` and all others were `0` (not sure what to call this, maybe a "single-entry vector").    The question I wanted to explore was what would be the maximum possible cosine similarities when comparing those vectors against the user's interest vector.

Maybe that's not the best approach though.  For one thing, I think you're right that consumers might want to specify more complex vectors for their content -- maybe we give them an easy way to specify a single-entry vector but also a way to specify all components.

Based on that, maybe there are some other ways to measure the strength of the interest detection:
  - Measure the cosine similarity between the user's interest vector and the "non-unknown vector" (constructed by using `1` for all interests, except unknown which is `0`.
  - Measure the maximum cosine similarity that content could possibly have (I think you can get this by copying the raw counts from the user's interest vector, but setting the unknown element to `0`).

Are there other ways to measure this?  How do you think we should approach it?

> There are a couple of ways to normalize a category hit vector into scores, such as arithmetic means or [soft-max](https://en.wikipedia.org/wiki/Softmax_function).

Thanks!  I'll look into those and probably switch to using one of those.

Reply:
Ah, I see. You're already thinking about relevance metrics!

Yup, i think we call those "single-entry vector" or "unit vector". Likewise, [one-hot encoding](https://en.wikipedia.org/wiki/One-hot) is also popular to convert categorical data into numerical vectors.

There are also many other similarity metrics which do not need encoding, such as [Jaccard_index](https://en.wikipedia.org/wiki/Jaccard_index). All in all, we have lots of options (including the cosine one). I think we can pick a few to examine and experiment with.

> This comes from an assumption that consumers would only want to specify a single interest for a suggestion, so we would have a vector where one element was 1 and all others were 0.

It depends. For suggest, yes, it's more common to assign one category to each suggestion type. Whereas for other content such as Pocket stories on New Tab, they normally come with more than 1 category. So we might want to design for more generic cases for similarity calculation.

Reply:
I'm going to need to do some more research on all of this, but I updated `metrics.yaml` with my current plan:
- Calculate 3 measures of cosine similarity, by comparing the user's interest vector with:
   - The single-entry vector that's closest to the user's interest vector
   - The multiple-entry vector that's closest to the user's interest vector (but with `0` for the inconclusive component)
   - The "non-inconclusive" vector
- Multiply by 100, to make it an integer for Glean, but don't try to normalize.   We can figure that out later.